pkg/skaffold/jib/jib.go

```diff
@@ -17,64 +17,172 @@ limitations under the License.
+var watchedFiles = sync.Map{}
+
+// getDependencies returns a list of files to watch for changes to rebuild
+func getDependencies(cmd *exec.Cmd, projectName string) ([]string, error) {
+	var dependencyList []string
+	f, ok := watchedFiles.Load(projectName)
+	if !ok {
+		f = filesLists{}
+	}
+	files := f.(filesLists)
+
+	template := files.PathMap
+	if len(template.Inputs) == 0 && len(template.Build) == 0 {
+		// Make sure build file modification time map is setup
+		if files.BuildFileTimes == nil {
+			files.BuildFileTimes = make(map[string]time.Time)
+		}
+
+		// Refresh dependency list if empty
+		if err := refreshDependencyList(&files, cmd); err != nil {
+			return nil, errors.Wrap(err, "initial Jib dependency refresh failed")
+		}
+
+	} else if err := walkFiles(&template.Build, &template.Ignore, func(path string, info os.FileInfo) error {
+		// Walk build files to check for changes
+		if val, ok := files.BuildFileTimes[path]; !ok || info.ModTime() != val {
+			return refreshDependencyList(&files, cmd)
+		}
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to walk Jib build files for changes")
+	}
+
+	// Walk updated files to build dependency list
+	if err := walkFiles(&files.PathMap.Inputs, &files.PathMap.Ignore, func(path string, info os.FileInfo) error {
+		dependencyList = append(dependencyList, path)
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to walk Jib input files to build dependency list")
+	}
+	if err := walkFiles(&files.PathMap.Build, &files.PathMap.Ignore, func(path string, info os.FileInfo) error {
+		dependencyList = append(dependencyList, path)
+		files.BuildFileTimes[path] = info.ModTime()
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to walk Jib build files to build dependency list")
 	}
 
-	cmdDirInfo, err := os.Stat(cmd.Dir)
+	// Store updated files list information
+	watchedFiles.Store(projectName, files)
+
+	sort.Strings(dependencyList)
+	return dependencyList, nil
+}
+
+// refreshDependencyList calls out to Jib to update files.PathMap with the latest list of files/directories to watch.
+func refreshDependencyList(files *filesLists, cmd *exec.Cmd) error {
+	stdout, err := util.RunCmdOut(cmd)
 	if err != nil {
-		return nil, err
+		return errors.Wrap(err, "failed to get Jib dependencies; it's possible you are using an old version of Jib (Skaffold requires Jib v1.0.2+)")
 	}
 
-	// Parses stdout for the dependencies, one per line
+	// Search for Jib's output JSON. Jib's Maven/Gradle output takes the following form:
+	// ...
+	// BEGIN JIB JSON
+	// {"build":["/paths","/to","/buildFiles"],"inputs":["/paths","/to","/inputs"],"ignore":["/paths","/to","/ignore"]}
+	// ...
+	// To parse the output, search for "BEGIN JIB JSON", then unmarshal the next line into the pathMap struct.
 	lines := util.NonEmptyLines(stdout)
+	for i := range lines {
+		if lines[i] == "BEGIN JIB JSON" {
+			line := strings.Replace(lines[i+1], "\\", "\\\\", -1)
```

Comment:
this spooks me a little bit, can you add a small comment explaining why you're doing this?